trinity/sync/common/chain.py

```diff
@@ -124,6 +123,29 @@ def register_peer(self, peer: BasePeer) -> None:
                 else:
                     self.run_task(self.sync(peer))
 
+    @property
+    def _syncing(self) -> bool:
+        return self._peer_header_syncer is not None
+
+    @contextmanager
+    def _get_peer_header_syncer(
+            self, peer: HeaderRequestingPeer) -> Generator['PeerHeaderSyncer', None, None]:
+        if self._syncing:
+            raise ValidationError("Cannot sync headers from two peers at the same time")
+
+        self._peer_header_syncer = PeerHeaderSyncer(
+            self.chain,
+            self.db,
+            peer,
+            self.header_queue,
```

Comment:
Just noting that in my opinion, much of the confusing complexity we've gotten rid of involved passing around queues.  If you are onboard with avoiding this pattern it doesn't need to be part of this PR but I my thoughts on what it might look like are:

Instead of the service being responsible for both generating headers and feeding them into the queue, potentially we should move the responsibility into the service where the `header_queue` lives.  This could be a background task/process/daemon which does roughly:

```
async def consume_headers(self, peer_header_syncer):
    async for header in peer_header_syncer.stream():
        self.header_queue.put_nowait(header)
```

Reply:
Note that this is a `TaskQueue` not an `asyncio.Queue`. A nice thing about using a `TaskQueue` here is that there is back-pressure when downstream processes are slowing down. (So we can `await self.header_queue.add(headers)` and it will pause if downstream hasn't finished)

I'll take a crack at removing this particular coupling, but I still think it could be a good solution for communicating between the `BaseHeaderChainSyncer` and, say, the `FastChainSyncer`. (which is how it is now, but they are implemented as superclass/subclass, so maybe it's less noticeable)

I'm also not remembering the queue complexity that you're talking about, so if you have a link/story about it, that would help me understand what to avoid.